File: algorithms/D/Delete Duplicate Folders in System/Delete Duplicate Folders in System.js
class Node{
    constructor(val,parent){
        this.val=val,this.parent=parent,this.children={},this.hash
    }
}
class Trie{
    constructor(){
        this.root=new Node('')
        this.hashMemo={}
    }
    insert(path){
        let cur=this.root
        for(let node of path )
            if(cur.children[node]!==undefined)
                cur=cur.children[node]
            else
                cur.children[node]=new Node(node,cur),
                cur=cur.children[node]
    }
    traverse(node=this.root){
        node.hash=[]
        if(node.children)
            for(let child of Object.values(node.children))
                node.hash.push(this.traverse(child))
        node.hash=node.hash.join('~')
        if(node.hash!==''){
            if(this.hashMemo[node.hash]===undefined)
                this.hashMemo[node.hash]=[] 
            this.hashMemo[node.hash].push(node)
            return node.val+'/['+node.hash+']'
        }
        return node.val
    }
    delete(node=this.root){
        if(!node || node.hash===undefined)
            return
        else if(this.hashMemo[node.hash] && this.hashMemo[node.hash].length>1)
            for(let todelete of this.hashMemo[node.hash])
                delete todelete.parent.children[todelete.val]
        else
            for(let child of Object.values(node.children))
                this.delete(child)
    }
    serialize(node=this.root,stack=[],res=[]){
        if(node.val!=='')
            stack.push(node.val),
            res.push([...stack])
        for(let child of Object.values(node.children))
            this.serialize(child,stack,res)
        stack.pop()
        return res
    }
}

var deleteDuplicateFolder = function(paths) {
    let T=new Trie()
    for(let P of paths)
        T.insert(P)
    T.traverse()
    T.delete()
    return T.serialize()
};

File: algorithms/D/Delete Duplicate Folders in System/Delete Duplicate Folders in System.cpp
class node{
    public: 
        map<string, node*> child;
        bool end;
        bool exclude;
        string name;
        node(string fname){
            exclude = false;
            end = false;
            name = fname;
        }
};

class Solution {
public:
    node *root;
    unordered_map<string, vector<node*>> parentMap;
    vector<vector<string>> ans;
    vector<vector<string>> deleteDuplicateFolder(vector<vector<string>>& paths) {
        root = new node("/");
        for(int i = 0 ; i < paths.size(); i++) insert(root, paths[i], 0);
        dfs(root);
        for(auto &e : parentMap)
            if(e.second.size() > 1)
                for(auto &parent : e.second)
                    parent->exclude = true;
        vector<string> temp;
        getans(root, temp);
        return ans;
    }
    
    void getans(node *root, vector<string> &temp){
        if(root->exclude) return;
        if(root->name != "/") temp.push_back(root->name);
        if(root->end) ans.push_back(temp);
        for(auto &e : root->child)
            getans(e.second, temp);
        if(root->name != "/") temp.pop_back();
    }
    
    string dfs(node *root){
        string childPath = "";
        for(auto &e : root->child){
            childPath += dfs(e.second);
            childPath += "|";
        }
        if(childPath != "") parentMap[childPath].push_back(root);
        return root->name + "|" + childPath;
    }
    
    void insert(node *root, vector<string> &path, int pos){
        if(pos < path.size()){
            if(root->child.count(path[pos]) == 0) root->child[path[pos]] = new node(path[pos]);
            insert(root->child[path[pos]], path, pos+1);
        }
        else if(pos == path.size()) root->end = true;
    }
};
